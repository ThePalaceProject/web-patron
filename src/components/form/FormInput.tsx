/** @jsxRuntime classic */
/** @jsx jsx */
// eslint-disable-next-line @typescript-eslint/ban-ts-comment
import { jsx } from "theme-ui";
import * as React from "react";
import FormLabel from "./FormLabel";
import TextInput from "../TextInput";

type FormInputProps = React.InputHTMLAttributes<HTMLInputElement> & {
  name: string;
  label: string;
  error?: string;
  required?: boolean;
  endIcon?: React.ReactNode;
};

const END_ICON_WIDTH = 40;
/**
 * An input component to be used in a form that supports errors,
 * labels and uses the base TextInput component
 */
const FormInput = React.forwardRef<HTMLInputElement, FormInputProps>(
  ({ type = "text", label, name, error, required, endIcon, ...props }, ref) => {
    const describedById = `${name}-errors`;
    return (
      <div role="group" sx={{ mb: 2 }}>
        <FormLabel sx={{ mb: 1, display: "inline-block" }} htmlFor={name}>
          {label}
          {required && <span sx={{ color: "ui.error" }}>*</span>}
        </FormLabel>
        <div sx={{ position: "relative" }}>
          <TextInput
            id={name}
            name={name}
            aria-label={`${name} input`}
            aria-describedby={describedById}
            ref={ref}
            type={type}
            sx={{
              display: "block",
              borderColor: error ? "ui.error" : undefined,
              paddingRight: endIcon ? END_ICON_WIDTH : undefined
            }}
            {...props}
          />
          {endIcon && (
            <div
              sx={{
                width: END_ICON_WIDTH,
                position: "absolute",
                right: 0,
                top: 0,
                height: "100%"
              }}
            >
              {endIcon}
            </div>
          )}
        </div>

        <span
          id={describedById}
          sx={{ color: "ui.error", fontStyle: "italic" }}
        >
          {error}
        </span>
      </div>
    );
  }
);

export default FormInput;
