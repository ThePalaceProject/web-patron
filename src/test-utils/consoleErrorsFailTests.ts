import { jest } from "@jest/globals";
const consoleErrorSpy = jest.spyOn(global.console, "error");

// use this to make console errors fail tests. Useful when running in to issues.
export default function consoleErrorsFailTests() {
  consoleErrorSpy.mockImplementation((msg, ...opts) => {
    console.warn(msg, ...opts);
    throw new Error(msg);
  });
}
