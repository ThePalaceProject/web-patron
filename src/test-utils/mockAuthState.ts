import { AuthCredentials, OPDS1 } from "interfaces";
import Cookie from "js-cookie";
import { generateCredentials } from "utils/auth";
import { basicTokenAuthenticationUrl } from "./fixtures";

const mockCookie = Cookie as any;

export const creds: AuthCredentials = {
  token: "some-token",
  methodType: OPDS1.BasicAuthType
};

export const persistentUserCredentials = generateCredentials("1234", "pinpin");
export const firstToken = "IaMaBeArErToKeN";
export const tokenCreds1: AuthCredentials = {
  token: {
    basicToken: persistentUserCredentials,
    bearerToken: firstToken
  },
  authenticationUrl: basicTokenAuthenticationUrl,
  methodType: OPDS1.BasicTokenAuthType
};

export const newToken = "IaMaBeArErToKeN2";
export const tokenCreds2: AuthCredentials = {
  token: {
    basicToken: persistentUserCredentials,
    bearerToken: newToken
  },
  authenticationUrl: basicTokenAuthenticationUrl,
  methodType: OPDS1.BasicTokenAuthType
};

const str = JSON.stringify;

export default function mockAuthenticatedOnce(
  credentials: AuthCredentials | null = creds
) {
  mockCookie.get.mockReturnValueOnce(str(credentials));
}

export function mockAuthenticated(credentials: AuthCredentials | null = creds) {
  mockCookie.get.mockReturnValue(str(credentials));
}
