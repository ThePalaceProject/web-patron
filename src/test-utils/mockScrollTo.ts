import { jest } from "@jest/globals";
/**
 * JSdom doesn't include a stub for the scrollTo function, which we use inside of
 * Lane. Therefore we will stub it globally here
 */

Element.prototype.scrollTo = jest.fn();
window.scrollTo = jest.fn();

export default {};
