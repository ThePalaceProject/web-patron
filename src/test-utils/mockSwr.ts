import { jest } from "@jest/globals";
import { responseInterface } from "swr";

export function makeSwrResponse<T>(value?: Partial<responseInterface<T, any>>) {
  return {
    data: undefined,
    error: undefined,
    revalidate: jest.fn(),
    isValidating: false,
    mutate: jest.fn(),
    ...value
  };
}

export type MockSwr<T> = (value?: Partial<responseInterface<T, any>>) => void;
