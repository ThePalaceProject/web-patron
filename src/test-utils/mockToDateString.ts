import { jest } from "@jest/globals";
/**
 * This file mocks Date.prototype.toDateString so that it just returns a
 * stub string. This makes code using that function testable across timezones
 * since we can't reliably set the timezone of a node process to something specific.
 */
export const MOCK_DATE_STRING = "mock-date-string";
export const toDateStringSpy = jest
  .spyOn(Date.prototype, "toDateString")
  .mockImplementation(() => MOCK_DATE_STRING);
