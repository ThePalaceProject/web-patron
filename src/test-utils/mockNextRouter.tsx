// eslint-disable-next-line @typescript-eslint/ban-ts-comment
// @ts-nocheck
import { jest } from "@jest/globals";
import * as React from "react";
import Router from "next/router";
import { RouterContext } from "next/dist/shared/lib/router-context.shared-runtime";
import { NextRouter } from "next/router";
import { libraryData } from "test-utils/fixtures";

/**
 * Mock for the next/Router import.
 */

export const mockPush = jest.fn().mockImplementation(async () => true);
Router.push = mockPush;

export const MockNextRouterContextProvider: React.FC<{
  router?: Partial<NextRouter>;
}> = ({ router = {}, children }) => {
  const {
    isReady = true,
    basePath = "",
    route = "",
    pathname = "",
    query = {
      // add the default library slug to the default query params
      library: libraryData.slug
    },
    // default as path is the home page
    asPath = `/${libraryData.slug}`,
    push = mockPush,
    replace = jest.fn().mockImplementation(async () => true),
    reload = jest.fn().mockImplementation(() => null),
    back = jest.fn().mockImplementation(() => null),
    prefetch = jest.fn().mockImplementation(async () => undefined),
    beforePopState = jest.fn().mockImplementation(() => null),
    isFallback = false,
    isLocaleDomain = false,
    isPreview = false,
    events = {
      on: () => null,
      off: () => null,
      emit: () => null
    }
  } = router;
  return (
    <RouterContext.Provider
      value={{
        isReady,
        basePath,
        route,
        pathname,
        query,
        asPath,
        push,
        replace,
        reload,
        back,
        prefetch,
        beforePopState,
        isFallback,
        isLocaleDomain,
        isPreview,
        events
      }}
    >
      {children}
    </RouterContext.Provider>
  );
};

export default Router;
