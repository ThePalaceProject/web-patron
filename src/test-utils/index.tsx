import * as React from "react";
import { render } from "@testing-library/react";
import "@testing-library/jest-dom";
import { beforeEach, expect, jest } from "@jest/globals";
import Adapter from "enzyme-adapter-react-16";
import { configure } from "enzyme";
import { LibraryData } from "../interfaces";
import "./mockScrollTo";
import * as fixtures from "./fixtures";
import serializer from "jest-emotion";
import { Provider as ReakitProvider } from "reakit";
import { MockNextRouterContextProvider } from "./mockNextRouter";
import { NextRouter } from "next/router";
import { enableFetchMocks } from "jest-fetch-mock";
import { LibraryProvider } from "components/context/LibraryContext";
import { BreadcrumbProvider } from "components/context/BreadcrumbContext";
import { UserContext, UserState } from "components/context/UserContext";
import { ThemeProvider } from "theme-ui";
import makeTheme from "theme";
import mockConfig from "test-utils/mockConfig";
import track from "analytics/track";
import "react-intersection-observer/test-utils";
import "test-utils/mockToDateString";

enableFetchMocks();
expect.addSnapshotSerializer(serializer);

// standard config mock
beforeEach(() => {
  mockConfig();
});

// mock the error tracker to prevent unnecessary console logs
export const mockTrackError = jest.fn();
track.error = mockTrackError;

export { fixtures };

// configure the enzyme adapter
configure({ adapter: new Adapter() });

/**
 * mock out the window.URL.createObjectURL since it isn't
 * available on jsdom
 */
const mockCreateObjectURL = jest.fn();
(global as any)["URL"].createObjectURL = mockCreateObjectURL;

type CustomRenderOptions = Parameters<typeof render>[1] & {
  router?: Partial<NextRouter>;
  library?: Partial<LibraryData>;
  user?: Partial<UserState>;
};
/**
 * Our custom render function wraps components in mocked versions of our
 * providers
 */
const customRender = (ui: any, options?: CustomRenderOptions) => {
  const library: LibraryData = {
    ...fixtures.libraryData,
    ...options?.library
  };
  const theme = makeTheme(library.colors);

  const user: UserState = {
    ...fixtures.user,
    ...options?.user
  };

  const AllTheProviders: React.FC = ({ children }) => {
    return (
      <MockNextRouterContextProvider router={options?.router}>
        <ThemeProvider theme={theme}>
          <ReakitProvider>
            <LibraryProvider library={library}>
              <UserContext.Provider value={user}>
                <BreadcrumbProvider>{children}</BreadcrumbProvider>
              </UserContext.Provider>
            </LibraryProvider>
          </ReakitProvider>
        </ThemeProvider>
      </MockNextRouterContextProvider>
    );
  };
  return {
    ...render(ui, { wrapper: AllTheProviders, ...options })
  };
};

// re-export everything
export * from "@testing-library/react";

// override render method
export { customRender as render };
