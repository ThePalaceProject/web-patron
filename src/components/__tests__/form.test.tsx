import * as React from "react";
import { render } from "../../test-utils";
import FormInput from "../form/FormInput";

test("renders as expected", () => {
  const utils = render(<FormInput name="input-name" label="Email" />);
  expect(utils.container.firstChild).toMatchSnapshot();
});

test("displays star when required", () => {
  const utils = render(<FormInput required name="input-name" label="Email" />);
  expect(utils.getByText("*")).toBeInTheDocument();
});

test("no star when not required", () => {
  const utils = render(<FormInput name="input-name" label="Email" />);
  expect(utils.queryByText("*")).toBeNull();
});

test("displays label", () => {
  const utils = render(<FormInput name="input-name" label="Email" />);
  expect(utils.getByText("Email")).toBeInTheDocument();
});

test("shows error when present", () => {
  const utils = render(
    <FormInput name="input-name" label="Email" error="Something went wrong" />
  );
  expect(utils.getByText("Something went wrong")).toBeInTheDocument();
});
