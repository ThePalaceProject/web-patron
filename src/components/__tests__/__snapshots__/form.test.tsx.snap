// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders as expected 1`] = `
.emotion-2 {
  margin-bottom: 8px;
}

.emotion-0 {
  font-size: 0.875rem;
  font-weight: 500;
  margin-right: 8px;
  margin-bottom: 4px;
  display: inline-block;
}

.emotion-1 {
  border-radius: 2px;
  border: 1px solid #bdbdbd;
  padding-left: 8px;
  padding-right: 8px;
  padding-top: 4px;
  padding-bottom: 4px;
  font-size: 1rem;
  font-family: 'Open Sans',Helvetica,Arial,sans-serif;
  font-weight: 300;
  line-height: 1.5;
  width: 100%;
  display: block;
}

.emotion-1::-webkit-input-placeholder {
  font-style: italic;
  color: var(--theme-ui-colors-ui-gray-dark,#616161);
  font-size: 0.875rem;
}

.emotion-1::-moz-placeholder {
  font-style: italic;
  color: var(--theme-ui-colors-ui-gray-dark,#616161);
  font-size: 0.875rem;
}

.emotion-1:-ms-input-placeholder {
  font-style: italic;
  color: var(--theme-ui-colors-ui-gray-dark,#616161);
  font-size: 0.875rem;
}

.emotion-1::placeholder {
  font-style: italic;
  color: var(--theme-ui-colors-ui-gray-dark,#616161);
  font-size: 0.875rem;
}

<div
  class="emotion-2"
  role="group"
>
  <label
    class="emotion-0"
    for="input-name"
  >
    Email
  </label>
  <input
    aria-describedby="input-name-errors"
    aria-label="input-name input"
    class="emotion-1"
    id="input-name"
    name="input-name"
    type="text"
  />
</div>
`;
